Add zerohack xdr

WhizhackDevelopers opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=22750) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=22750) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=22750) it.</sub>

WhizhackDevelopers reviewed (commented):
fixed error for unreachable code

ostolero commented:
Hi @WhizhackDevelopers, Thank you for your contribution and patience in waiting for this review. I'm sorry for the delay in response as we have quite a few contributions to review at the moment.
Will get to it as soon as I can and update you with any required changes.
Thank you again.

ostolero commented on Packs/Zerohack_XDR/Integrations/ZerohackXDR/ZerohackXDR.py:
```diff
@@ -0,0 +1,331 @@
+    first_fetch_timestamp = int(first_fetch_time.timestamp()) if first_fetch_time else None
+
+    # First fetch timestamp failsafe.
+    assert isinstance(first_fetch_timestamp, int)
```
Let's throw an indicative exception here-
try:
 int(first_fetch) 
except TypeError: 
 raise DemistoException("the first fetch value is invalid")

ostolero commented on Packs/Zerohack_XDR/Integrations/ZerohackXDR/ZerohackXDR.py:
```diff
@@ -0,0 +1,331 @@
+                        'severity': convert_to_demisto_severity(alert.get('ids_threat_severity', 'Low')),
+                    }
+                    incidents.append(incident)
+                    last_incident_time = incident_created_time
```
As you fetch incidents for every severity seperatly, a better implimentatuion would be to save the last_fetch_time for every severity. see this scenario-
We are fetching maximum of 4 incidents from severity level 1 and 2.
We fetch 2 incidents from level 1 with creation time - 10:30 , 10:35 our last_incident_time = 10:35
There is another incidents from level 1 at 10:37 but as we are limited to 2 incidents we didn't fetch it yet.
We fetch 2 incidents from level 2 with creation time - 10:30, 10:40 our last_incident_time = 10:40

In our next fetch we will start from 10:40 missing our level 1 incident that occurred at 10:37.

A suggesion for a fix could be to save our last_incident_time per severity in a dictionary or list. then comapre to our last_run object and update the times by severity.
Let me know if you need any help with the implimentation :)

WhizhackDevelopers replied:
Your suggestions were insightful in improving the integration.

ostolero commented on Packs/Zerohack_XDR/Integrations/ZerohackXDR/README.md:
```diff
@@ -0,0 +1,2 @@
+Note: This is a beta pack, which lets you implement and test pre-release software. Since the pack is beta, it might contain bugs. Updates to the pack during the beta phase might include non-backward compatible features. We appreciate your feedback on the quality and usability of the pack to help us identify issues, fix them, and continually improve.
```
I don't see the beta flag in your yml file. If you want to release this pack as beta add beta:true to the yml, if not remove this comment

ostolero commented on Packs/Zerohack_XDR/Integrations/ZerohackXDR/README.md:
```diff
@@ -0,0 +1,2 @@
+Note: This is a beta pack, which lets you implement and test pre-release software. Since the pack is beta, it might contain bugs. Updates to the pack during the beta phase might include non-backward compatible features. We appreciate your feedback on the quality and usability of the pack to help us identify issues, fix them, and continually improve.
+
```
Please create a proper README file for the integration. You can generate on easily using our SDK command generate-docs https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/generate_docs/README.md

ostolero commented on Packs/Zerohack_XDR/Integrations/ZerohackXDR/ZerohackXDR.py:
```diff
@@ -0,0 +1,331 @@
+
+
+from typing import Any, Dict, Tuple, List, Optional, cast
```
add these imports - 
import demistomock as demisto
from CommonServerPython import *
from CommonServerUserPython import *

ostolero commented on Packs/Zerohack_XDR/Integrations/ZerohackXDR/ZerohackXDR.py:
```diff
@@ -0,0 +1,331 @@
+
```
add these imports - 
```
import demistomock as demisto
from CommonServerPython import *
from CommonServerUserPython import *
```

ostolero reviewed (changes requested):
Hi @WhizhackDevelopers , Thank you for the contribution excellent work!
I added a few comments for you to check out, mainly-
1. Change the way we handle last_run update, see a detailed comment in the function.
2. Generate a README file for the integration.
Also please check the validation errors generated in the build (mainly formatting issues). let me know if you have any questions on how to fix them :)
I'm asking our security team to review the playbooks here as well and he will update with any other required changes.
Thank you! 
and again I'm here for any questions here or via slack.

ostolero commented:
Hi @WhizhackDevelopers , The changes look great! I'm still waiting on a cr from @efelmandar, he will update here with any changes required soon.

efelmandar commented:
Hi @WhizhackDevelopers, Thank you for your contribution and patience as I reviewd the content.
The pack looks good but I have a couple of question:


1. I tried to upload the new incident-type file to XSOAR but got an error saying that the format is wrong, maybe you placed the wrong file? or the wrong content?

2. As I reviewd the playbook `Zerohack XDR_Sophos Firewall_Block IP` and from my understanding the purpose of this playbook is blocking an IP address in the sophos firewall, but the logic is identical to the existing `Sophos_Firewall_-_Block_IP` playbook in the SophosXGFirewall pack. if you want to use it with your own logic I would recommend adding it as a subplaybook or if there is another reason for why it is necessary in your pack as well, please explain and we'll discuss it as we try to reduce content duplication as possibe.

other than those issues the content looks good and as soon as we'll resolve them we will be able to go ahead with the merge of the PR

WhizhackDevelopers commented:
Hi @efelmandar  i have reviewed the issue mentioned 
1. There was formatting issue in the IncidentType-Zerohack_XDR.json which i have fixed
2.  The playbook provided is just a demo on how to use Sophos firewall with our integration, we would like your recommendation on what should be done since it is not necessary .

WhizhackDevelopers commented:
> Hi @WhizhackDevelopers , Thank you for the contribution excellent work! I added a few comments for you to check out, mainly-
> 
>     1. Change the way we handle last_run update, see a detailed comment in the function.
> 
>     2. Generate a README file for the integration.
>        Also please check the validation errors generated in the build (mainly formatting issues). let me know if you have any questions on how to fix them :)
>        I'm asking our security team to review the playbooks here as well and he will update with any other required changes.
>        Thank you!
>        and again I'm here for any questions here or via slack.

I have commented regarding the second  query and fixed the first query please check and revert

ostolero reviewed (approved):
Hi @WhizhackDevelopers the changes are on point, just need @efelmandar final aprroval and we can move to demo :)

ostolero commented:
Hi @WhizhackDevelopers , in the meantime you can check the errors in the build and try to fix them, let me know if you have any questions I'm available here and on slack 👍

WhizhackDevelopers commented:
> Hi @WhizhackDevelopers , in the meantime you can check the errors in the build and try to fix them, let me know if you have any questions I'm available here and on slack +1

Hi @ostolero  the automated test errors are trivial which can be ignored , also i was trying to connect on slack but was not able to get the invitation to work .

efelmandar commented:
Hi @WhizhackDevelopers, I went over the changes in the `IncidentType-Zerohack_XDR.json` file, but the schema is still not valid as an XSOAR incident type. from what I can tell, the data in this file looks like maybe it is a json of an incident data and if so it should go into the `test_data` directory.   
In the `IncidentTypes` folder (and I'll assume this content is for XSOAR since there is no `marketplaces` field in pack_metadata) should be only incident type content items ([link to docs](https://xsoar.pan.dev/docs/incidents/incident-types#incident-type-properties)) which I recommend exporting from the UI once it is fully configured.

In addition and to relate it to the playbook, when configuring the `Incident type` we can set a default playbook and set if it should run automatically or not and set a special layout (or leave it to use the default one), in the `indicators extraction` tab you can set `what to extract` to `extract all indicators from all fields` which will auto-extract and auto-enrich indicators after fetching the incident. 
As for the playbook it mostly depends on the incident data from your system but for a basic playbook to run I would suggest creating a playbook which maybe check the incident severity and check enrichment for any extracted indicators and add the `Block IP - Generic v3` as a sub-playbook and pass it the IP address to block which will block it in any supported system the user might have - including `Sophos Firewall - block IP` as you previously did.

If this is not needed in your use-case design you can have no incident type and let the user configure a default incident type from the existing one for any fetched incident - but this is less recommended.
**see screenshots below**

**Incident Type example**
![image](https://user-images.githubusercontent.com/102903097/210239046-3d051af9-6c88-4afc-a14f-723da0b5b8c5.png)

**Incident type creation**
![image](https://user-images.githubusercontent.com/102903097/210238443-bcb5335c-ad04-4add-a353-71396537bd93.png)

WhizhackDevelopers commented:
> Hi @WhizhackDevelopers, I went over the changes in the `IncidentType-Zerohack_XDR.json` file, but the schema is still not valid as an XSOAR incident type. from what I can tell, the data in this file looks like maybe it is a json of an incident data and if so it should go into the `test_data` directory. In the `IncidentTypes` folder (and I'll assume this content is for XSOAR since there is no `marketplaces` field in pack_metadata) should be only incident type content items ([link to docs](https://xsoar.pan.dev/docs/incidents/incident-types#incident-type-properties)) which I recommend exporting from the UI once it is fully configured.
> 
> In addition and to relate it to the playbook, when configuring the `Incident type` we can set a default playbook and set if it should run automatically or not and set a special layout (or leave it to use the default one), in the `indicators extraction` tab you can set `what to extract` to `extract all indicators from all fields` which will auto-extract and auto-enrich indicators after fetching the incident. As for the playbook it mostly depends on the incident data from your system but for a basic playbook to run I would suggest creating a playbook which maybe check the incident severity and check enrichment for any extracted indicators and add the `Block IP - Generic v3` as a sub-playbook and pass it the IP address to block which will block it in any supported system the user might have - including `Sophos Firewall - block IP` as you previously did.
> 
> If this is not needed in your use-case design you can have no incident type and let the user configure a default incident type from the existing one for any fetched incident - but this is less recommended. **see screenshots below**
> 
> **Incident Type example** ![image](https://user-images.githubusercontent.com/102903097/210239046-3d051af9-6c88-4afc-a14f-723da0b5b8c5.png)
> 
> **Incident type creation** ![image](https://user-images.githubusercontent.com/102903097/210238443-bcb5335c-ad04-4add-a353-71396537bd93.png)

I have rectified my mistake from [IncidentType-Zerohack_XDR.json]  and pulled it  from the UI 
I have also removed the playbook directory since the playbook was same as Sophos Firewall_Block IP with the input as IP pulled from the integration.  
Please review the changes and update on any issues

efelmandar commented:
@WhizhackDevelopers Great work fixing the incident type, now it looks good. I made a small change and change it to use default playbook for now, since you removed the playbook from the pack.
@ostolero @WhizhackDevelopers please let me know if and when will be a good time for a short demo?

ostolero commented:
Hi @WhizhackDevelopers ,
I tried to fix as many validations as possible but the ones that are left require your attention.
I see the `Create ID set` fails which looks like a bug, I'll check it and get back to you.
But we will not be able to merge without all the steps of the build passing.
I'll be happy to help with any questions about how to fix these.

esharf commented on Packs/Zerohack_XDR/IncidentTypes/IncidentType-Zerohack_XDR.json:
```diff
@@ -0,0 +1,44 @@
+[
```
```suggestion
```

esharf commented on Packs/Zerohack_XDR/IncidentTypes/IncidentType-Zerohack_XDR.json:
```diff
@@ -0,0 +1,44 @@
+			"fieldCliNameToExtractSettings": {}
+		}
+	}
+]
```
```suggestion
```

esharf reviewed (commented):
should probably fix the ID set creation error

ostolero commented:
Hi @WhizhackDevelopers , Do you need help with the validation errors? is there another reason this is pending? If you are not working on this anymore we can close this and continue when you come back to it.

WhizhackDevelopers commented:
> Hi @WhizhackDevelopers , Do you need help with the validation errors? is there another reason this is pending? If you are not working on this anymore we can close this and continue when you come back to it.

hello @ostolero  i need help with validation errors is there a way to connect except slack , i tried to connect on slack but was not  able to get the invitation to work.

ostolero commented:
Hi @WhizhackDevelopers , I sent you an email so we can connect via zoom and go over the issues.
We can continue the discussion over email.

ostolero commented:
Hi @WhizhackDevelopers , great work fixing all the validation errors! we can go ahead and schedule our demo.
Let's set up a time through slack.
